mne/encoding/feature.py

```diff
@@ -0,0 +1,319 @@
+    decim : int
+        Decimate by choosing only a subsample of data points. Highly
+        recommended for data recorded at high sampling frequencies, as
+        otherwise huge intermediate matrices have to be created and inverted.
+    tstep : float
+        Length of windows for peak-to-peak detection for raw data cleaning.
+    nonzero : bool
+        Whether to reject timepoints with nonzero values in all input features.
+        Defaults to True.
+
+    Returns
+    -------
+    X_cleaned : array, shape (n_clean_times, n_features)
+        The input X array w/ non-active and noisy rows removed.
+    y_cleaned : array, shape (n_clean_times, n_channels)
+        The input y array w/ non-active and noisy rows removed.
+    """
+
+    # find only those positions where at least one predictor isn't 0
+    if X.shape[0] != y.shape[0]:
+        raise ValueError('X and y have different numbers of timepoints')
+    if isinstance(X, sparse.spmatrix):
+        nonzero_rows = np.unique(X.nonzero()[0])
+        is_sparse = True
+    else:
+        nonzero_rows = np.unique(np.where((X != 0).any(1))[0])
+        is_sparse = False
+    keep_rows = nonzero_rows if nonzero is True else np.arange(X.shape[0])
+
+    # reject positions based on extreme steps in the data
+    if reject is not None:
+        _, inds_bad = _reject_data_segments(y.T, reject, flat, decim=None,
+                                            info=info, tstep=tstep)
+        # Expand to include all bad indices, and remove from rows to keep
+        inds_bad = np.hstack(range(t0, t1) for t0, t1 in inds_bad)
+        keep_rows = np.setdiff1d(keep_rows, inds_bad)
+
+    if is_sparse is True:
+        X = X.tocsr()[keep_rows]
+    else:
+        X = X[keep_rows]
+
+    return X, y[keep_rows]
+
+
+def _delay_timeseries(ts, delays, sfreq=1.):
+    """Return a time-lagged input timeseries.
+
+    Parameters
+    ----------
+    ts: array, shape (n_times, n_feats)
+        The timeseries to delay
+    delays: array of floats, shape (n_delays,)
+        The time (in seconds) of each delay for specifying
+        pre-defined delays. Negative means timepoints in the past,
+        positive means timepoints in the future.
+    sfreq: int
+        The sampling frequency of the series
+
+    Returns
+    -------
+    delayed: array, shape(n_times, n_feats * n_delays)
+        The delayed matrix
+    """
+    n_times, n_feats = ts.shape
+    delays_ixs = np.round((delays * sfreq)).astype(int)
+
+    # Remove duplicated ixs
+    if np.unique(delays_ixs).shape[0] != delays_ixs.shape[0]:
+        warn('Converting delays to indices resulted in duplicates.')
+
+    # Convert ts to dense if it is sparse
+    if isinstance(ts, sparse.spmatrix):
+        is_sparse = True
+        sp_type = type(ts)
+        ts = ts.todense()
+    else:
+        is_sparse = False
+
+    # Iterate through indices and append
+    delayed = []
+    delayed_names = []
```

Comment:
For sure - it'd be pretty easy to just do this before the pipeline as well...just put it in here to show what kinds of stuff one could do w/ the pipeline.